File: java/src/main/java/com/bookstore/PromotedBook.java
package com.bookstore;

/**
 * Concrete books on promotion have several goodies: <br>
 * - You get a 10% discount on your next purchase.<br>
 * - You get a discount on delivery if it is heavy.<br>
 * - You get a gift wrapping for free.<br>
 */
public class PromotedBook extends AbstractItemOnOffer {

    public PromotedBook(String name) {
        this(name, new BookWeight(name));
    }

    private PromotedBook(String name, Weight weight) {
        super(name, weight, new DeliveryCostLines( // 
                new DeliveryCostPromotionHeavyWeight(weight) //
        ));
    }

    @Override
    public void putIntoMyCart(Cart cart) {
        cart.add(getName());
        new DiscountVoucher(10).putIntoMyCart(cart);
    }

    @Override
    protected void handleGiftOptions(Cart cart) {
        new GiftWrapping(this).putIntoMyCart(cart);
    }

}

File: cpp/book_store-cmake_catch2/src/PromotedBook.cpp
//
// Created by mille on 28-Sep-22.
//

#include "PromotedBook.h"
#include "DiscountVoucher.h"
#include "GiftWrapping.h"

namespace bookstore {

  int PromotedBook::DEFAULT_WEIGHT_BOOK_GRAMMS = 500;
  PromotedBook::PromotedBook(const std::string& _name) : AbstractItemOnOffer(_name) { }

  void PromotedBook::putIntoMyCart(Cart& cart) {
    cart.add(getName());
    DiscountVoucher(10).putIntoMyCart(cart);
  }

  int PromotedBook::itemWeight() {
    if (getName().find("C++") != std::string::npos) {
      return 750;
    }
    else if (getName().find("Good Parts") != std::string::npos) {
      return 100;
    }
    return DEFAULT_WEIGHT_BOOK_GRAMMS;
  }

  bool PromotedBook::hasDiscountOnDelivery() {
    return itemWeight() >= DEFAULT_WEIGHT_BOOK_GRAMMS;
  }

  void PromotedBook::handleGiftOptions(Cart& cart) {
    GiftWrapping(*this).putIntoMyCart(cart);
  }
} // bookstore